File: plugin/flat_histogram/tutorial/tutorial_8_trappe.py
import argparse
import feasst as fst

print(fst.version())
parser = argparse.ArgumentParser()
parser.add_argument("--task", type=int, help="SLURM job array index", default=0)
parser.add_argument("--num_procs", type=int, help="number of processors", default=12)
parser.add_argument("--num_hours", type=float, help="number of hours before restart", default=5*24)
parser.add_argument("--dccb_cutoff", type=int, help="cutoff for dual cut configurational bias", default=4)
parser.add_argument("--dccb_begin", type=int, help="number of molecules before DCCB", default=300)
parser.add_argument("--lx", type=float, help="box length in x", default=33.0)
parser.add_argument("--ly", type=float, help="box length in y", default=33.0)
parser.add_argument("--lz", type=float, help="box length in z", default=33.0)
parser.add_argument("--max_particles", type=int, help="maximum number of particles", default=400)
parser.add_argument("--temperature", type=float, help="temperature in Kelvin", default=263.15)
parser.add_argument("--beta_mu", type=float, help="chemical potential times inverse temperature", default=0.)
parser.add_argument("--particles", "-p", help="particle types", action='append', required=True)
parser.add_argument("--collect_flatness", type=int, help="number of WL flatness to begin collection", default=18)
parser.add_argument("--min_flatness", type=int, help="number of WL flatness to switch to TM", default=22)
args = parser.parse_args()
print("args:", args)

def mc(thread, mn, mx):
    steps_per=int(1e4)
    mc = fst.MakeMonteCarlo()
    domain_args = {"side_length0": str(args.lx), "side_length1": str(args.ly), "side_length2": str(args.lz)}
    if mx > args.dccb_begin:
        domain_args["init_cells"] = str(args.dccb_cutoff)
        ref = "0"
        num = "4"
    else:
        ref = "-1"
        num = "1"
    config_args = dict()
    index = 0
    beta = 1./args.temperature
    criteria_args = {"beta": str(beta)}
    for part in args.particles:
        config_args["particle_type"+str(index)] = part
        criteria_args["chemical_potential"+str(index)] = str(args.beta_mu/beta)
        index += 1
    mc.add(fst.Configuration(fst.MakeDomain(fst.args(domain_args)), config_args))
    mc.add(fst.Potential(fst.MakeLennardJones()))
    mc.add(fst.Potential(fst.MakeLongRangeCorrections()))
    if mx > args.dccb_begin:
        if mc.configuration().domain().num_cells() > 0:
            reference = fst.Potential(fst.MakeLennardJones(), fst.MakeVisitModelCell())
        else:
            reference = fst.Potential(fst.MakeLennardJones())
        reference.set_model_params(mc.configuration())
        for site_type in range(mc.configuration().num_site_types()):
            reference.set_model_param("cutoff", site_type, args.dccb_cutoff)
        mc.add_to_reference(reference)
    mc.add(fst.MakeFlatHistogram(
        fst.MakeMacrostateNumParticles(
            fst.Histogram(fst.args({"width": "1", "max": str(mx), "min": str(mn)}))),
        # fst.MakeTransitionMatrix(fst.args({"min_sweeps": str(args.sweeps)})),
        fst.MakeWLTM(fst.args({
            "collect_flatness": str(args.collect_flatness),
            "min_flatness": str(args.min_flatness),
            "min_sweeps": "1000"})),
        criteria_args))
    for particle_type in range(mc.configuration().num_particle_types()):
        mc.add(fst.MakeTrialTranslate(fst.args({
            "particle_type": str(particle_type),
            "weight": "1.",
            "tunable_param": "1.",
            "reference_index": ref,
            "num_steps": num})))
        mc.add(fst.MakeTrialRotate(fst.args({
            "particle_type": str(particle_type),
            "weight": "1.",
            "tunable_param": "1.",
            "reference_index": ref,
            "num_steps": num})))
        mc.add(fst.MakeTrialTransfer(fst.args({
            "particle_type": str(particle_type),
            "weight": "4",
            "reference_index": ref,
            "num_steps": num})))
    mc.add(fst.MakeCheckEnergy(fst.args({"steps_per": str(steps_per), "tolerance": "0.0001"})))
    mc.add(fst.MakeTuner(fst.args({"steps_per": str(steps_per), "stop_after_phase": "0"})))
    mc.add(fst.MakeLogAndMovie(fst.args({"steps_per": str(steps_per),
                                         "file_name": "clones" + str(thread),
                                         "file_name_append_phase": "True"})))
    if mc.configuration().num_particle_types() > 1:
        mc.add(fst.MakeNumParticles(fst.args({
          "particle_type": "0",
          "file_name": "num" + str(thread),
          "file_name_append_phase": "True",
          "steps_per_write": str(steps_per),
          "steps_per_update": "1",
          "multistate": "True"})))
    mc.add(fst.MakeCriteriaUpdater(fst.args({"steps_per": str(steps_per)})))
    mc.add(fst.MakeCriteriaWriter(fst.args({"steps_per": str(steps_per),
                                            "file_name": "clones" + str(thread) + "_crit.txt",
                                            "file_name_append_phase": "True"})))
    mc.set(fst.MakeCheckpoint(fst.args({"file_name": "checkpoint" + str(thread) + ".fst",
                                        "num_hours_terminate": str(0.9*args.num_procs*args.num_hours)})))
    mc.add(fst.MakeAnalyzeRigidBonds(fst.args({"steps_per": str(steps_per)})))
    return mc

windows=fst.WindowExponential(fst.args({
    "alpha": "1.75",
    "num": str(args.num_procs),
    "maximum": str(args.max_particles),
    "extra_overlap": "2"})).boundaries()
print(windows)

if args.task == 0:
    clones = fst.MakeClones()
    for proc, win in enumerate(windows):
        clones.add(mc(proc, win[0], win[1]))
    clones.set(fst.MakeCheckpoint(fst.args({"file_name": "checkpoint.fst"})))
else:
    clones = fst.MakeClones("checkpoint", args.num_procs);
#clones.initialize_and_run_until_complete()
clones.initialize_and_run_until_complete(fst.args({"ln_prob_file": "ln_prob.txt"}))
print(clones.ln_prob().values())
open('clones.fst', 'w').write(clones.serialize())

File: plugin/flat_histogram/tutorial/tutorial_8_trappe.cpp
#include <assert.h>
#include <fstream>
#include "feasst.h"

feasst::ArgumentParse args("A grand canonical ensemble flat histogram Monte Carlo simulation of a bulk fluid", {
  {"--task", "SLURM job array index", "0"},
  {"--num_procs", "number of processors", "12"},
  {"--num_hours", "number of hours before restart", feasst::str(5*24)},
  {"--dccb_cutoff", "cutoff for dual cut configurational bias", "4"},
  {"--dccb_begin", "number of molecules before using DCCB", "300"},
  {"--lx", "box length in x", "33.0"},
  {"--ly", "box length in y", "33.0"},
  {"--lz", "box length in z", "33.0"},
  {"--min_particles", "minimum number of particles", "0"},
  {"--max_particles", "maximum number of particles", "400"},
  {"--temperature", "temperature in Kelvin", "263.15"},
  {"--particle0", "particle type 0"},
  {"--particle1", "particle type 1, not used if empty (the default)"},
  {"--collect_flatness", "number of WL flatness to begin collection", "18"},
  {"--min_flatness", "number of WL flatness to switch to TM", "22"},
  {"--beta_mu", "baseline chemical potential of each species", "-7"},
  {"--delta_beta_mu1", "beta_mu1 = beta_mu0 + delta_beta_mu1", "0."},
  {"--cyl_radius", "radius of cylindrical confinement, not used if < 0", "-1"},
  {"--cyl_cutoff", "square well interaction distance from cylinder to point", "10"},
  {"--cyl_epsilon", "square well interaction strength", "500"},
  {"--file_xyz", "file name of xyz positions of initial configuration. "
    "Assume all particle0, and do not use if empty (the default)"}});

std::shared_ptr<feasst::MonteCarlo> mc(const int thread, const int mn, const int mx) {
  const std::string steps_per = feasst::str(1e5);
  auto mc = feasst::MakeMonteCarlo();
  feasst::argtype domain_args = {
    {"side_length0", args.get("--lx")},
    {"side_length1", args.get("--ly")},
    {"side_length2", args.get("--lz")}};
  const int dccb_begin = args.get_int("--dccb_begin");
  std::string ref("-1");
  std::string num("1");
  const int dccb_cutoff = args.get_int("--dccb_cutoff");
  if (mx > dccb_begin) {
    domain_args.insert({"init_cells", feasst::str(dccb_cutoff)});
    ref = "0";
    num = "4";
  }
  const double beta = 1./args.get_double("--temperature");
  const double beta_mu = args.get_double("--beta_mu");
  feasst::argtype config_args = {{"particle_type0", args.get("--particle0")}};
  feasst::argtype criteria_args = {{"beta", feasst::str(beta)},
    {"chemical_potential0", feasst::str(beta_mu/beta)}};
  if (args.option_given("--particle1")) {
    config_args.insert({"particle_type1", args.get("--particle1")});
    criteria_args.insert({"chemical_potential1", feasst::str((beta_mu + args.get_double("--delta_beta_mu1"))/beta)});
  }
  if (args.option_given("--file_xyz")) {
    assert(thread == 0); // cannot initialize multiple threads to same configuration
    feasst::Configuration config(feasst::MakeDomain(domain_args), config_args);
    feasst::FileXYZ().load(args.get("--file_xyz"), &config);
    mc->add(config);
  } else {
    mc->add(feasst::Configuration(feasst::MakeDomain(domain_args), config_args));
  }
  mc->add(feasst::Potential(feasst::MakeLennardJones()));
  mc->add(feasst::Potential(feasst::MakeLongRangeCorrections()));
  const double cyl_radius = args.get_double("--cyl_radius");
  if (cyl_radius > 0) {
    feasst::Potential cylinder(feasst::MakeModelSquareWellShape(feasst::MakeCylinder(
      {{"radius", feasst::str(cyl_radius)}},
      feasst::Position({{"x", "0"}, {"y", "0"}, {"z", "0"}}),
      feasst::Position({{"x", "0"}, {"y", "0"}, {"z", "1"}}))));
    cylinder.set_model_params(mc->configuration());
    for (int site_type = 0; site_type < mc->configuration().num_site_types(); ++site_type) {
      cylinder.set_model_param("cutoff", site_type, args.get_double("--cyl_cutoff"));
      cylinder.set_model_param("epsilon", site_type, args.get_double("--cyl_epsilon"));
    }
    mc->add(cylinder);
  }
  if (mx > dccb_begin) {
    feasst::Potential reference(feasst::MakeLennardJones());
    if (mc->configuration().domain().num_cells() > 0) {
      reference = feasst::Potential(feasst::MakeLennardJones(),
                                    feasst::MakeVisitModelCell({{"min_length", feasst::str(dccb_cutoff)}}));
    }
    reference.set_model_params(mc->configuration());
    for (int site_type = 0; site_type < mc->configuration().num_site_types(); ++site_type) {
      reference.set_model_param("cutoff", site_type, dccb_cutoff);
    }
    mc->add_to_reference(reference);
  }
  mc->set(feasst::MakeThermoParams(criteria_args));
  mc->set(feasst::MakeFlatHistogram(
    feasst::MakeMacrostateNumParticles(
      feasst::Histogram({{"width", "1"}, {"max", feasst::str(mx)}, {"min", feasst::str(mn)}})),
    // feasst::MakeTransitionMatrix(feasst::args({"min_sweeps": feasst::str(args.sweeps)})),
    feasst::MakeWLTM({
      {"collect_flatness", args.get("--collect_flatness")},
      {"min_flatness", args.get("--min_flatness")},
      {"min_sweeps", "1000"}})));
  std::cout << "Initial energy: " << MAX_PRECISION << mc->criteria().current_energy() << std::endl;
  for (int particle_type = 0; particle_type < mc->configuration().num_particle_types(); ++particle_type) {
    mc->add(feasst::MakeTrialTranslate({{"particle_type", feasst::str(particle_type)}, {"weight", "1."},
      {"tunable_param", "1."}, {"reference_index", ref}, {"num_steps", num}}));
    if (mx > dccb_begin && mc->configuration().particle_type(particle_type).num_sites() == 2) {
      mc->add(feasst::MakeTrialGrow({
        {{"transfer", "true"},  // {"regrow", "true"},  # regrow isn't very efficient
         {"particle_type", feasst::str(particle_type)}, {"site", "0"}, {"weight", "4"}},
        {{"bond", "true"}, {"mobile_site", "1"}, {"anchor_site", "0"}}},
        {{"reference_index", ref}, {"num_steps", num}}));
      mc->add(feasst::MakeTrialGrow({
        {{"particle_type", feasst::str(particle_type)}, {"weight", "0.5"},
         {"bond", "true"}, {"mobile_site", "1"}, {"anchor_site", "0"},
         {"reference_index", ref}, {"num_steps", num}}}));
      mc->add(feasst::MakeTrialGrow({
        {{"particle_type", feasst::str(particle_type)}, {"weight", "0.5"},
         {"bond", "true"}, {"mobile_site", "0"}, {"anchor_site", "1"},
         {"reference_index", ref}, {"num_steps", num}}}));
    } else {
      mc->add(feasst::MakeTrialRotate({{"particle_type", feasst::str(particle_type)}, {"weight", "1."},
        {"tunable_param", "1."}, {"reference_index", ref}, {"num_steps", num}}));
      mc->add(feasst::MakeTrialTransfer({{"particle_type", feasst::str(particle_type)},
        {"weight", "4"}, {"reference_index", ref}, {"num_steps", num}}));
    }
  }
  mc->add(feasst::MakeCheckEnergy({{"steps_per", steps_per}, {"tolerance", "0.0001"}}));
  mc->add(feasst::MakeTuner({{"steps_per", steps_per}, {"stop_after_phase", "0"}}));
  mc->add(feasst::MakeLogAndMovie({{"steps_per", steps_per},
                                   {"file_name", "clones" + feasst::str(thread)},
                                   {"file_name_append_phase", "True"}}));
  mc->add(feasst::MakeEnergy({
    {"file_name", "en" + feasst::str(thread) + ".txt"},
    {"file_name_append_phase", "true"},
    {"start_after_phase", "0"},
    {"steps_per_write", steps_per},
    {"steps_per_update", "1"},
    {"multistate", "True"}}));
  if (mc->configuration().num_particle_types() > 1) {
    mc->add(feasst::MakeNumParticles({
      {"particle_type", "0"},
      {"file_name", "num" + feasst::str(thread) + ".txt"},
      {"file_name_append_phase", "true"},
      {"start_after_phase", "0"},
      {"steps_per_write", steps_per},
      {"steps_per_update", "1"},
      {"multistate", "True"}}));
  }
  mc->add(feasst::MakeExtensiveMoments({
    {"steps_per_write", steps_per},
    {"file_name", "extmom"+feasst::str(thread) + ".txt"},
    {"file_name_append_phase", "true"},
    {"start_after_phase", "0"}, // do not update until equilibration complete (phase 1)
    {"max_order", "3"},
    {"multistate", "True"}}));
  mc->add(feasst::MakeCriteriaUpdater({{"steps_per", steps_per}}));
  mc->add(feasst::MakeCriteriaWriter({{"steps_per", steps_per},
                                     {"file_name", "clones" + feasst::str(thread) + "_crit.txt"},
                                     {"file_name_append_phase", "True"}}));
  mc->set(feasst::MakeCheckpoint({{"file_name", "checkpoint" + feasst::str(thread) + ".fst"},
                                 {"num_hours_terminate", feasst::str(0.99*args.get_int("--num_procs")*args.get_double("--num_hours"))}}));
  mc->add(feasst::MakeCheckRigidBonds({{"steps_per", steps_per}}));
  return mc;
}

int main(int argc, char ** argv) {
  std::cout << feasst::version() << std::endl;
  std::cout << "args: " << args.parse(argc, argv) << std::endl;
  const int max_particles = args.get_int("--max_particles");
  const int min_particles = args.get_int("--min_particles");
  const int num_procs = args.get_int("--num_procs");
  auto windows = feasst::WindowExponential({
    {"alpha", "1.75"},
    {"num", feasst::str(num_procs)},
    {"minimum", feasst::str(min_particles)},
    {"maximum", feasst::str(max_particles)},
    {"extra_overlap", "2"}}).boundaries();
  std::cout << feasst::feasst_str(windows) << std::endl;
  auto clones = feasst::MakeClones();
  if (args.get_int("--task") == 0) {
    for (int proc = 0; proc < static_cast<int>(windows.size()); ++proc) {
      clones->add(mc(proc, windows[proc][0], windows[proc][1]));
    }
    clones->set(feasst::MakeCheckpoint({{"file_name", "checkpoint.fst"}}));
  } else {
    clones = feasst::MakeClones("checkpoint", num_procs);
  }
  clones->initialize_and_run_until_complete({{"ln_prob_file", "ln_prob.txt"}});
  std::cout << feasst::feasst_str(clones->ln_prob().values()) << std::endl;
  std::ofstream file("clones.fst");
  file << clones->serialize();
  file.close();
}
